Killing complex impl for CloseableThreadLocal / DisposableThreadLocal…

ayende opened the pull request:
… in favor of an impl that doesn't use underlying ThreadLocal

ayende committed:
RavenDB-14555 killing complex impl for CloseableThreadLocal / DisposableThreadLocal in favor of an impl that doesn't use underlying ThreadLocal

arekpalinski commented on src/Lucene.Net/Util/LightThreadLocal.cs:
```diff
@@ -0,0 +1,94 @@
+            private readonly HashSet<LightThreadLocal<T>> _parents = new HashSet<LightThreadLocal<T>>(ReferenceComparer<LightThreadLocal<T>>.Instance);
+            public void Register(LightThreadLocal<T> parent)
+            {
+                _parents.Add(parent);
```
This is allowed to be run concurrently, isn't it? Then shouldn't we use `ConcurrentBag` instead of `HashSet`?

arekpalinski replied:
Hm.. but now I see it needs to ensure uniqueness. Sill - is it safe to use regular `HashSet` here?

ayende replied:
This is thread static value, no concurrency possible

ppekrol commented on src/Lucene.Net/Util/LightThreadLocal.cs:
```diff
@@ -0,0 +1,144 @@
+            }
+        }
+        public void Dispose()
+        {
```
We are missing `GC.SuppressFinalize(this);` here

ayende replied:
Good catch, yes

ppekrol replied:
I've fixed that

ppekrol committed:
RavenDB-14555 prevent double disposal

ayende committed:
RavenDB-14555
- Making sure that we can clear local thread set when the ThreadLocal is disposed.
- Fixing test
